### PR TITLE
is787/invitation-registration

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Allows registration only by invitation
- option configurable at app.login
- invitations are generated with confirmation-codes.
- confirmation codes have one use and have a lifetime of 5 days (currently hard-coded)
- ui and backend
- added restAPI entrypoint to retrieve front-end config

## Related issue number

Implements #787 , #791 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] **Runs in the swarm**
- [x] If you design a new module, add your user to .github/CODEOWNERS

### CODE REVIEW

File: api/specs/webserver/v0/components/schemas/registration.yaml [hunk 1]
```diff
@@ -26,7 +26,16 @@ RegistrationType:
...
+  required:
+    - email
+    - password
+
+
   example:
     email: foo@mymail.com
     password: 'my secret'
     confim: 'my secret'
+    type: 'ABCDE1234'
```
Comment (sanderegg) [inline api/specs/webserver/v0/components/schemas/registration.yaml]: should be invation no?

File: services/web/client/source/class/qxapp/io/rest/ResourceFactory.js [hunk 2]
```diff
@@ -25,6 +25,45 @@ qx.Class.define("qxapp.io.rest.ResourceFactory", {
...
+    __config: null,
+
+    requestConfig: function() {
+      if (this.__config === null) {
+        // SEE: https://www.qooxdoo.org/current/pages/communication/rest.html
+        // SEE: api/specs/webserver/v0/openapi.yaml
+        const basePath = qxapp.io.rest.ResourceFactory.API;
+        // Singular resource
+        const configCheck = new qxapp.io.rest.Resource({
+          // Get health check
```
Comment (sanderegg) [inline services/web/client/source/class/qxapp/io/rest/ResourceFactory.js]: get config... or remove comment...

File: services/web/server/src/simcore_service_webserver/login/handlers.py [hunk 3]
```diff
@@ -30,19 +31,24 @@
...
 async def register(request: web.Request):
     _, _, body = await extract_and_validate(request)
 
     # see https://aiohttp.readthedocs.io/en/stable/web_advanced.html#data-sharing-aka-no-singletons-please
+    app_cfg = get_login_config(request.app) # TODO: replace cfg by app_cfg
     db = get_storage(request.app)
+
     email = body.email
-    username = email.split('@')[0]
+    username = email.split('@')[0] # FIXME: this has to be unique and add this in user registration!
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/handlers.py]: why unique? the name of the user??

File: services/web/server/src/simcore_service_webserver/login/registration.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    )
+    return confirmation
+
+
+def get_confirmation_info(confirmation):
+    info = confirmation
+
+    # data
+    try:
+        info['data'] = json.loads(confirmation['data'])
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/registration.py]: what is the purpose of this? why would json loads fail?

File: services/web/server/src/simcore_service_webserver/login/registration.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    return info
+
+
+def get_invitation_url(confirmation, origin: URL=None) -> URL:
+    code = confirmation['code']
+    assert confirmation['action'] == ConfirmationAction.INVITATION.name
+
+    if origin is None:
+        origin = URL()
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/login/registration.py]: why should it be empty?

File: services/web/server/src/simcore_service_webserver/login/handlers.py (hunk 3 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/login/handlers.py]: This is not human-readable email. It is like in github. The username is a short name used as identifier and unique in the system (e.g. in your case there is a unique ``sanderegg`` in github so I can e.g. refer to you in markdown as ``@sanderegg``)

File: services/web/server/src/simcore_service_webserver/login/registration.py (hunk 5 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/login/registration.py]: These are helper functions. By default it returns just the path and not the origin

File: services/web/server/src/simcore_service_webserver/login/registration.py (hunk 4 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/login/registration.py]: The column "data" is a defined in the db model as a string for the moment. Therefore the format that is injected is NOT validated and therefore not guaranteed

**Pull request merged**